sklearn/utils/_param_validation.py

```diff
@@ -425,21 +425,117 @@ def generate_invalid_param_val(constraint):
+
+    constraints : list of Constraint
+        The list of all constraints for this parameter.
+
+    Returns
+    -------
+    val : object
+        A value that does not satisfy the interval constraint.
+    """
+    if interval.left is None and interval.right is None:
         raise NotImplementedError
+
+    if interval.type is Real:
+        # generate a non-integer value such that it can't be valid even if there's also
+        # an integer interval constraint.
+        if interval.left is not None:
+            return np.floor(interval.left) - 0.5
+        else:  # right is not None
+            return np.ceil(interval.right) + 0.5
+
+    else:  # interval.type is Integral
+        # We need to check if there's also a real interval constraint to generate a
+        # value that is not valid for any of the 2 interval constraints.
+        real_intervals = [
+            i for i in constraints if isinstance(i, Interval) and i.type is Real
+        ]
+        real_interval = real_intervals[0] if real_intervals else None
+
+        if real_interval is None:
+            # Only the integer interval constraint -> easy
+            if interval.left is not None:
+                return interval.left - 1
+            else:
+                return interval.right + 1
+
+        # There's also a real interval constraint. Try to find a value left to both or
+        # right to both or in between them.
+
+        # redefine left and right bounds to be smallest and largest valid integers in
+        # both intervals.
+        int_left = interval.left
+        if int_left is not None and interval.closed in ("right", "neither"):
+            int_left = int_left + 1
+
+        int_right = interval.right
+        if int_right is not None and interval.closed in ("left", "neither"):
+            int_right = int_right - 1
+
+        real_left = real_interval.left
+        if real_interval.left is not None:
+            real_left = int(np.ceil(real_interval.left))
+            if real_interval.closed in ("right", "neither"):
+                real_left = real_left + 1
+
+        real_right = real_interval.right
+        if real_interval.right is not None:
+            real_right = int(np.floor(real_interval.right))
+            if real_interval.closed in ("left", "neither"):
+                real_right = real_right - 1
+
+        if int_left is not None and real_left is not None:
+            # there exists an int left to both intervals
+            return min(int_left, real_left) - 1
+
+        if int_right is not None and real_right is not None:
+            # there exists an int right to both intervals
+            return max(int_right, real_right) + 1
+
+        if int_left is not None:
+            if real_right is not None and int_left - real_right >= 2:
+                # there exists an int between the 2 intervals
+                return int_left - 1
+            else:
+                raise NotImplementedError
```

Comment:
It could be nice to have an additional error message to explain in which case we are just in case that we wrongly use the helper.

Reply:
Then we could add a check in the test for this case.

Reply:
it's not that the helper is wrongly used. It means that there exists no invalid value for the constraint. For instance, all lists are valid for the constraint _InstancesOF(list) so you can't generate a list that does not satisfy the constraint.

I added a comment to explain that, and a test to cover this (although already covered by the common test)